direct/nn/varsplitnet/varsplitnet.py

```diff
@@ -0,0 +1,211 @@
+# coding=utf-8
+# Copyright (c) DIRECT Contributors
+
+from typing import Callable, Optional
+
+import torch
+from torch import nn
+
+from direct.data.transforms import expand_operator, reduce_operator
+from direct.nn.build_nn_model import ModelName, _build_model
+
+
+class MRIVarSplitNet(nn.Module):
+    r"""MRI reconstruction network that solves the variable split optimisation problem.
+
+    It solves the following:
+
+    .. math ::
+         z^{i-1} = \arg \min_{z} \mu * ||x^{i-1} - z||_2^2 + \mathcal{R}(z)
+
+         x^{i} = \arg \min_{x} ||y - A(x)||_2^2 + \mu * ||x - z^{i-1}||_2^2
+
+    by unrolling twice using the gradient descent algorithm and replacing :math:`R` with a neural network.
+    More specifically, for :math:`z_0, x_0 = \text{SENSE}(\tilde{y})`:
+
+    .. math ::
+        z^{i} = \alpha_{i-1} \times f_{\theta_{i-1}}\Big(\mu(z^{i-1} - x^{i-1}), z^{i-1}\Big), \quad i=1,\cdots,T_{reg}
+
+    where :math:`x^{i}` is the output of
+
+    .. math ::
+        (x^{i})^{j} = (x^{i})^{j-1} - \beta_{j-1} \Big[ A^{*}\big( A( (x^{i})^{j-1} ) - \tilde{y} \big) +
+        \mu ((x^{i})^{j-1} - z^{i}) \Big], \quad j=1,\cdots,T_{dc},
+
+    i.e. :math:`x^{i}=(x^{i}^{T_{reg}})`.
+    """
+
+    def __init__(
+        self,
+        forward_operator: Callable,
+        backward_operator: Callable,
+        num_steps_reg: int,
+        num_steps_dc: int,
+        image_init: str = "sense",
+        no_parameter_sharing: bool = True,
+        image_model_architecture: ModelName = ModelName.unet,
+        kspace_no_parameter_sharing: Optional[bool] = True,
+        kspace_model_architecture: Optional[ModelName] = None,
+        **kwargs,
+    ):
+        """Inits :class:`MRIVarSplitNet`."""
+        super().__init__()
+        self.num_steps_reg = num_steps_reg
+        self.num_steps_dc = num_steps_dc
+
+        self.image_nets = nn.ModuleList()
+        if kspace_model_architecture:
+            self.kspace_nets = nn.ModuleList()
+        else:
+            self.kspace_nets = None
+
+        self.no_parameter_sharing = no_parameter_sharing
+
+        assert image_model_architecture in ["unet", "normunet", "resnet", "didn", "conv"]
+        assert kspace_model_architecture in ["unet", "normunet", "resnet", "didn", "conv", None]
```

Comment:
No asserts in non-testing code

Reply:
Do you think things like this should be done with an if statement throwing some for of invalid input exception instead? Or not at all?

In this case I do think it would be better solved with some proper typing of the input :)